### ISSUE TITLE
Config system: bug when combining context manager and direct value setting

### ISSUE THREAD

**Issue by jakevdp** [google/jax#9304]
This is the cause of the issue discussed in https://github.com/google/jax/pull/9301#issuecomment-1020592833
```python
import jax

print(jax.config.jax_numpy_rank_promotion)  # allow
with jax.numpy_rank_promotion('warn'):
  print(jax.config.jax_numpy_rank_promotion)  # warn
  jax.config.FLAGS.jax_numpy_rank_promotion = 'raise'
  print(jax.config.jax_numpy_rank_promotion)  # warn
print(jax.config.jax_numpy_rank_promotion)  # raise
```
I believe the last line should print `allow`, but changing the FLAG state within the context manager is confusing. This pattern is used in a test here: https://github.com/google/jax/blob/b8372b0ca25df235f230a1fff3030f420cac3248/tests/api_test.py#L3226-L3231

I'm not sure whether the test intended to exercise this behavior.

**Comment by hawkinsp** [comment]
I'm guessing that the test did intend to mix changing the local and global flag state.

**Comment by jakevdp** [comment]
OK, makes sense.

So this means that the test is incompatible with decorators like `jax.disable_rank_promotion('allow')`, because the decorator uses a thread-local context manager that is subject to this bug, causing a test failure if you decorate the test case.

**Comment by hawkinsp** [comment]
Well, the test might be buggy, but (absent threading) it's supposed to change the state locally, restoring it to the way it was at the end of the test. I would have expected it to pass under an outer context manager: context managers should nest correctly. Maybe the test is buggy in some way, though.

**Comment by jakevdp** [comment]
Yeah, I think the test is buggy. It looks like this:
```python
allow_promotion = config.jax_numpy_rank_promotion  # reads the thread-local state
try:
  ...
  FLAGS.jax_numpy_rank_promotion = 'raise'  # changes the flag state
finally:
  FLAGS.jax_numpy_rank_promotion = allow_promotion  # sets the flag state to the thread local state
```
The result is that the ending flag state may be different than the beginning flag state.

**Comment by jakevdp** [comment]
Maybe we could do it with the private `config._read` method

**Comment by jakevdp** [comment]
Another sign the test is buggy: it fails if you change the rank promotion flag:
```
$ JAX_NUMPY_RANK_PROMOTION=raise pytest tests/api_test.py::RankPromotionTest::test_rank_promotion_forces_retrace
<...>
============================================================================================ short test summary info =============================================================================================
FAILED tests/api_test.py::RankPromotionTest::test_rank_promotion_forces_retrace - AssertionError: 2 != 3
```

**Comment by jakevdp** [comment]
The test issue should be fixed in #9325

**Comment by jakevdp** [comment]
I think the behavior described originally in this issue makes sense: the context manager modifies the thread local state of the value, while setting the flag value effectively controls the default which is unmodified by changing the thread local state. I'm going to close this as I think it's working correctly.
